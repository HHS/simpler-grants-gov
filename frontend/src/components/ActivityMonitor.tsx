"use client";

import { useUser } from "src/services/auth/useUser";

import { useCallback, useEffect, useRef, useState } from "react";

// when a user is logged in, this will track all clicks and key presses and,
// on each action, if the user's auth token is expiring or expired, will refresh the
// token or log the user out accordingly
export function ActivityMonitor() {
  const { user, refreshIfExpiring, refreshIfExpired, logoutLocalUser } =
    useUser();

  const [listening, setListening] = useState(false);
  const handlerRef = useRef<EventListener>(null);

  const refreshTokenIfExpiringOrLogout = useCallback(() => {
    refreshIfExpired()
      .then((expired) => {
        if (expired) {
          logoutLocalUser();
          return;
        }
        return refreshIfExpiring();
      })
      .catch((e) => {
        console.error("Error refreshing token in response to user activity", e);
      });
  }, [refreshIfExpired, refreshIfExpiring, logoutLocalUser]);

  const addHandlers = useCallback(() => {
    if (!listening) {
      document.addEventListener("click", refreshTokenIfExpiringOrLogout);
      document.addEventListener("keydown", refreshTokenIfExpiringOrLogout);
      setListening(true);
    }
  }, [refreshTokenIfExpiringOrLogout, listening]);

  const removeHandlers = useCallback(() => {
    if (!listening) {
      return;
    }
    document.removeEventListener("click", handlerRef.current as EventListener);
    document.removeEventListener(
      "keydown",
      handlerRef.current as EventListener,
    );
    setListening(false);
    handlerRef.current = null;
  }, [listening]);

  // when logged in status changes, remove or add handlers
  useEffect(() => {
    if (!user || !user.token) {
      removeHandlers();
      return;
    }
    addHandlers();
    return () => {
      removeHandlers();
    };
  }, [user, addHandlers, removeHandlers]);

  // whenever we are not listening for activity, and the handler function has been updated
  // update the ref, so that we have the right reference ready when we want to remove the handlers later.
  // we do not want to pick up any changes made while listening, as those changes will not corerspond to the
  // handler that is actually attached to the document.
  // note that if we try to do this in addHandlers it somehow doesn't work right, given all of the dependencies involved.
  useEffect(() => {
    if (refreshTokenIfExpiringOrLogout && !listening) {
      handlerRef.current = refreshTokenIfExpiringOrLogout;
    }
  }, [refreshTokenIfExpiringOrLogout, listening]);

  return null;
}
