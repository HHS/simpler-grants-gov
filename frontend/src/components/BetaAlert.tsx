import { useTranslations } from "next-intl";
import { ReactNode } from "react";
import { GridContainer, Alert as USWDSAlert } from "@trussworks/react-uswds";

type BetaAlertProps = {
  containerClasses?: string;
  heading?: string;
  alertMessage?: string | ReactNode;
};

const BetaAlert = ({
  containerClasses,
  heading,
  alertMessage,
}: BetaAlertProps) => {
  const t = useTranslations("BetaAlert");
  const defaultAlertMessage = t.rich("alert", {
    LinkToGrants: (content) => <a href="https://www.grants.gov">{content}</a>,
  });

  return (
    <div data-testid="beta-alert" className={containerClasses}>
      <GridContainer>
        <USWDSAlert
          type="warning"
          headingLevel="h2"
          heading={heading || t("alertTitle")}
          noIcon
        >
          {alertMessage || defaultAlertMessage}
        </USWDSAlert>
      </GridContainer>
    </div>
  );
};

export default BetaAlert;
