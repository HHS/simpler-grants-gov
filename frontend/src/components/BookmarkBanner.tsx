import { useTranslations } from "next-intl";
import { GridContainer, Alert as USWDSAlert } from "@trussworks/react-uswds";

type BookmarkBannerProps = {
  containerClasses?: string;
  message?: string;
};

const BookmarkBanner = ({ containerClasses, message }: BookmarkBannerProps) => {
  const t = useTranslations("BookmarkBanner");

  const defaultMessage = (
    <>
      {t("message")}
      <br />
      {t.rich("technicalSupportMessage", {
        mailToGrants: (content) => (
          <a href="mailto:simpler@grants.gov">{content}</a>
        ),
      })}
    </>
  );

  return (
    <div data-testid="bookmark-banner" className={containerClasses}>
      <GridContainer>
        <USWDSAlert type="info" headingLevel="h2" heading={t("title")}>
          {message || defaultMessage}
        </USWDSAlert>
      </GridContainer>
    </div>
  );
};

export default BookmarkBanner;
