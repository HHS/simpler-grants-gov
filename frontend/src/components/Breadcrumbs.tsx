import {
  Breadcrumb,
  BreadcrumbBar,
  BreadcrumbLink,
  GridContainer,
} from "@trussworks/react-uswds";

export type Breadcrumb = {
  title: string;
  path: string;
};

export type BreadcrumbList = Breadcrumb[];

type Props = {
  breadcrumbList: BreadcrumbList;
};

const microdata = {
  ol: {
    itemScope: true,
    itemType: "https://schema.org/BreadcrumbList",
  },
  li: {
    itemScope: true,
    itemProp: "itemListElement",
    itemType: "https://schema.org/ListItem",
  },
  a: {
    itemScope: true,
    itemProp: "item",
    itemType: "https://schema.org/WebPage",
  },
};

const Breadcrumbs = ({ breadcrumbList }: Props) => {
  const breadcrumArray = breadcrumbList.map((breadcrumbInfo, i) => {
    return (
      <Breadcrumb
        key={breadcrumbInfo.title + "-crumb"}
        current={i + 1 === breadcrumbList.length}
        {...microdata.li}
      >
        {i + 1 !== breadcrumbList.length ? (
          <BreadcrumbLink
            href={breadcrumbInfo.path}
            itemID={breadcrumbInfo.path}
            {...microdata.a}
          >
            {}
            <span itemProp="name">{breadcrumbInfo.title}</span>
          </BreadcrumbLink>
        ) : (
          <span itemProp="name">{breadcrumbInfo.title}</span>
        )}
        <meta itemProp="position" content={(i + 1).toString()} />
      </Breadcrumb>
    );
  });

  return (
    <GridContainer data-testid="breadcrumb">
      <BreadcrumbBar listProps={{ ...microdata.ol }}>
        {breadcrumArray}
      </BreadcrumbBar>
    </GridContainer>
  );
};

export default Breadcrumbs;
