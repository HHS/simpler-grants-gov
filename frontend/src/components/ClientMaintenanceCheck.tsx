"use client";

import { useFeatureFlags } from "src/hooks/useFeatureFlags";

import { useEffect } from "react";

/**
 * Checks for the "major" features being offline, otherwise we're not in maintenance, so redir to homepage.
 */
export default function ClientMaintenanceCheck() {
  const {
    checkFeatureFlag, // An instance of FeatureFlagsManager
  } = useFeatureFlags();

  useEffect(() => {
    if (
      !checkFeatureFlag("searchOff") &&
      !checkFeatureFlag("opportunityOff") &&
      checkFeatureFlag("authOn")
    ) {
      // This piece does not seem reliable, maybe due to re-render or rehyrdrate cylces where we redirect away before the actual client live feature flag is set
      // redirect("/", RedirectType.push);
    }
  }, [checkFeatureFlag]);

  return null;
}
