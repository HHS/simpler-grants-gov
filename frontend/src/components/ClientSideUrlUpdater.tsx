"use client";

import { useSearchParamUpdater } from "src/hooks/useSearchParamUpdater";

import { useRouter } from "next/navigation";
import { useEffect } from "react";

// This should be used as a way to force a url change from a server component
// supports updating:
// - the full url
// - an individual query param
// - the query string
// Note that if you want to zero out the query or param value, you should explicitly pass "" (empty string)
export function ClientSideUrlUpdater({
  param,
  value,
  url,
  query,
}: {
  param?: string;
  value?: string;
  url?: string;
  query?: string;
}) {
  const router = useRouter();
  const { updateQueryParams, searchParams } = useSearchParamUpdater();
  // if query has been passed in, use that, otherwise pass through the current query
  const updatedQuery = query !== undefined ? query : searchParams.get("query");

  useEffect(() => {
    if (url) {
      router.push(url);
    }
  }, [url, router]);
  useEffect(() => {
    if (param && value !== undefined) {
      updateQueryParams(value, param, updatedQuery);
    }
  }, [param, value, updatedQuery, updateQueryParams]);
  return <></>;
}
