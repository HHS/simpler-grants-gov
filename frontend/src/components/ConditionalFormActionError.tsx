import { ErrorMessage } from "@trussworks/react-uswds";

export function ConditionalFormActionError<
  T extends { [key: string]: string[] },
>({ errors, fieldName }: { errors?: T; fieldName: keyof T }) {
  if (!errors) {
    return;
  }
  const error = errors[fieldName];
  if (error) {
    return <ErrorMessage className="maxw-mobile-lg">{error[0]}</ErrorMessage>;
  }
}
