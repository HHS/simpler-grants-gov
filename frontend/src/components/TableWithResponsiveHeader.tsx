import clsx from "clsx";

import { ReactNode } from "react";
import { Table } from "@trussworks/react-uswds";

export type TableCellData = {
  cellData: ReactNode;
  stackOrder?: number; // -1 to hide
  className?: string;
};

export const TableWithResponsiveHeader = ({
  headerContent,
  tableRowData,
}: {
  headerContent: TableCellData[];
  tableRowData: TableCellData[][];
}) => {
  if (
    !tableRowData.every((tableRow) => tableRow.length === headerContent.length)
  ) {
    console.error(
      "Header and data content have mismatched link, unable to display responsive table",
    );
    return;
  }

  const headerNodes = headerContent.map((headerItem, i) => {
    return (
      <th
        key={`responsiveHeaderItem-${i}`}
        scope="col"
        className={clsx(
          "bg-base-lightest padding-y-205 minw-15",
          headerItem.className,
        )}
      >
        {headerItem.cellData}
      </th>
    );
  });

  const dataRows = tableRowData.map((tableRow, i) => {
    return (
      <tr
        key={`responsiveTableRow-${i}`}
        className="border-base border-x border-y tablet-lg:border-0"
      >
        {tableRow.map((tableCell, j) => {
          const stackOrder = tableCell.stackOrder || 0;
          return (
            <td
              key={`responsiveTableCell-${i}-${j}`}
              className={clsx(
                "tablet-lg:display-table-cell",
                "border-base",
                `order-${stackOrder}`,
                `tablet-lg:order-${j}`,
                tableCell.className,
                {
                  "display-none": stackOrder < 0,
                  "display-block": stackOrder > -1,
                },
              )}
            >
              <div className="display-flex flex-align-center">
                <div
                  className="tablet-lg:display-none flex-1 text-bold"
                  data-testid={`responsive-header-${i}-${j}`}
                >
                  {headerContent[j].cellData}
                </div>
                <div
                  className="flex-2 tablet:flex-3"
                  data-testid={`responsive-data-${i}-${j}`}
                >
                  {tableCell.cellData}
                </div>
              </div>
            </td>
          );
        })}
      </tr>
    );
  });

  return (
    <Table className="simpler-responsive-table width-full tablet-lg:width-auto border-base tablet-lg:border-0">
      <thead>
        <tr>{headerNodes}</tr>
      </thead>
      <tbody>{dataRows}</tbody>
    </Table>
  );
};
