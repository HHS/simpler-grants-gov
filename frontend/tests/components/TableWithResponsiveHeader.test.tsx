import { render, screen } from "@testing-library/react";
import { axe } from "jest-axe";
import {
  fakeResponsiveTableHeaders,
  fakeResponsiveTableRows,
} from "src/utils/testing/fixtures";

import { TableWithResponsiveHeader } from "src/components/TableWithResponsiveHeader";

// this does not directly test responsive aspects of the component, that should be done in e2e tests
// see https://github.com/HHS/simpler-grants-gov/issues/5414
describe("TableWithResponsiveHeader", () => {
  it("passes accessibility test", async () => {
    const component = TableWithResponsiveHeader({
      headerContent: fakeResponsiveTableHeaders,
      tableRowData: fakeResponsiveTableRows,
    });
    const { container } = render(component);
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });
  it.skip("matches snapshot", () => {
    const component = TableWithResponsiveHeader({
      headerContent: fakeResponsiveTableHeaders,
      tableRowData: fakeResponsiveTableRows,
    });
    const { container } = render(component);
    expect(container).toMatchSnapshot();
  });
  it("displays header content in table header", () => {
    render(
      <TableWithResponsiveHeader
        headerContent={fakeResponsiveTableHeaders}
        tableRowData={fakeResponsiveTableRows}
      />,
    );

    expect(
      screen.getByRole("columnheader", { name: "hi" }),
    ).toBeInTheDocument();
    expect(
      screen.getByRole("columnheader", { name: "a heading" }),
    ).toBeInTheDocument();
    expect(
      screen.getByRole("columnheader", { name: "table header cell" }),
    ).toBeInTheDocument();
  });
  it("displays header content (hidden) in table body", () => {
    render(
      <TableWithResponsiveHeader
        headerContent={fakeResponsiveTableHeaders}
        tableRowData={fakeResponsiveTableRows}
      />,
    );

    expect(screen.getByTestId("responsive-header-0-0")).toHaveTextContent("hi");
    expect(screen.getByTestId("responsive-header-0-0")).toHaveClass(
      "tablet-lg:display-none",
    );
    expect(screen.getByTestId("responsive-header-0-1")).toHaveTextContent(
      "a heading",
    );
    expect(screen.getByTestId("responsive-header-0-1")).toHaveClass(
      "tablet-lg:display-none",
    );
    expect(screen.getByTestId("responsive-header-0-2")).toHaveTextContent(
      "table header cell",
    );
    expect(screen.getByTestId("responsive-header-0-2")).toHaveClass(
      "tablet-lg:display-none",
    );
  });
  it("displays table content in table body rows", () => {
    render(
      <TableWithResponsiveHeader
        headerContent={fakeResponsiveTableHeaders}
        tableRowData={fakeResponsiveTableRows}
      />,
    );

    expect(screen.getByTestId("responsive-data-0-0")).toHaveTextContent(
      "hi from row one",
    );
    expect(screen.getByTestId("responsive-data-0-1")).toHaveTextContent(
      "i am column two",
    );
    expect(screen.getByTestId("responsive-data-0-2")).toHaveTextContent(
      "some data",
    );

    expect(screen.getByTestId("responsive-data-1-0")).toHaveTextContent(
      "hi from row two",
    );
    expect(screen.getByTestId("responsive-data-1-1")).toHaveTextContent(
      "still column two",
    );
    expect(screen.getByTestId("responsive-data-1-2")).toHaveTextContent(
      "some more data",
    );

    expect(screen.getByTestId("responsive-data-2-0")).toHaveTextContent(
      "hi from row three",
    );
    expect(screen.getByTestId("responsive-data-2-1")).toHaveTextContent(
      "column two",
    );
    expect(screen.getByTestId("responsive-data-2-2")).toHaveTextContent(
      "even more data",
    );
  });
  it("hides content with negative stack order", () => {
    render(
      <TableWithResponsiveHeader
        headerContent={fakeResponsiveTableHeaders}
        tableRowData={fakeResponsiveTableRows}
      />,
    );

    expect(
      screen.getByRole("cell", { name: "table header cell some data" }),
    ).toHaveClass("display-none");
    expect(
      screen.getByRole("cell", { name: "table header cell some more data" }),
    ).toHaveClass("display-none");
    expect(
      screen.getByRole("cell", { name: "table header cell even more data" }),
    ).toHaveClass("display-none");
  });
  it("applies stack order when present", () => {
    render(
      <TableWithResponsiveHeader
        headerContent={fakeResponsiveTableHeaders}
        tableRowData={fakeResponsiveTableRows}
      />,
    );

    expect(
      screen.getByRole("cell", { name: "hi hi from row one" }),
    ).toHaveClass("order-1");
    expect(
      screen.getByRole("cell", { name: "a heading i am column two" }),
    ).toHaveClass("order-0");
  });
  it("does not render if all table rows do not have the same number of items as the headers", () => {
    render(
      <TableWithResponsiveHeader
        headerContent={fakeResponsiveTableHeaders.concat([
          { cellData: "superfluous" },
        ])}
        tableRowData={fakeResponsiveTableRows}
      />,
    );

    expect(screen.queryByRole("table")).not.toBeInTheDocument();
  });
});
