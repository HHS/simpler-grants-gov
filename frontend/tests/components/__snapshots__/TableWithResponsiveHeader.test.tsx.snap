// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableWithResponsiveHeader matches snapshot 1`] = `
<div>
  <table
    class="usa-table usa-table--borderless simpler-responsive-table width-full tablet-lg:width-auto border-base tablet-lg:border-0"
    data-testid="table"
  >
    <thead>
      <tr>
        <th
          class="bg-base-lightest padding-y-205 minw-15"
          scope="col"
        >
          hi
        </th>
        <th
          class="bg-base-lightest padding-y-205 minw-15"
          scope="col"
        >
          a heading
        </th>
        <th
          class="bg-base-lightest padding-y-205 minw-15"
          scope="col"
        >
          table header cell
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        class="border-base border-x border-y tablet-lg:border-0"
      >
        <td
          class="tablet-lg:display-table-cell border-base order-1 tablet-lg:order-0 display-block"
        >
          <div
            class="display-flex flex-align-center"
          >
            <div
              class="tablet-lg:display-none flex-1 text-bold"
              data-testid="responsive-header-0-0"
            >
              hi
            </div>
            <div
              class="flex-2"
              data-testid="responsive-data-0-0"
            >
              hi from row one
            </div>
          </div>
        </td>
        <td
          class="tablet-lg:display-table-cell border-base order-0 tablet-lg:order-1 display-block"
        >
          <div
            class="display-flex flex-align-center"
          >
            <div
              class="tablet-lg:display-none flex-1 text-bold"
              data-testid="responsive-header-0-1"
            >
              a heading
            </div>
            <div
              class="flex-2"
              data-testid="responsive-data-0-1"
            >
              i am column two
            </div>
          </div>
        </td>
        <td
          class="tablet-lg:display-table-cell border-base order--1 tablet-lg:order-2 display-none"
        >
          <div
            class="display-flex flex-align-center"
          >
            <div
              class="tablet-lg:display-none flex-1 text-bold"
              data-testid="responsive-header-0-2"
            >
              table header cell
            </div>
            <div
              class="flex-2"
              data-testid="responsive-data-0-2"
            >
              some data
            </div>
          </div>
        </td>
      </tr>
      <tr
        class="border-base border-x border-y tablet-lg:border-0"
      >
        <td
          class="tablet-lg:display-table-cell border-base order-1 tablet-lg:order-0 display-block"
        >
          <div
            class="display-flex flex-align-center"
          >
            <div
              class="tablet-lg:display-none flex-1 text-bold"
              data-testid="responsive-header-1-0"
            >
              hi
            </div>
            <div
              class="flex-2"
              data-testid="responsive-data-1-0"
            >
              hi from row two
            </div>
          </div>
        </td>
        <td
          class="tablet-lg:display-table-cell border-base order-0 tablet-lg:order-1 display-block"
        >
          <div
            class="display-flex flex-align-center"
          >
            <div
              class="tablet-lg:display-none flex-1 text-bold"
              data-testid="responsive-header-1-1"
            >
              a heading
            </div>
            <div
              class="flex-2"
              data-testid="responsive-data-1-1"
            >
              still column two
            </div>
          </div>
        </td>
        <td
          class="tablet-lg:display-table-cell border-base order--1 tablet-lg:order-2 display-none"
        >
          <div
            class="display-flex flex-align-center"
          >
            <div
              class="tablet-lg:display-none flex-1 text-bold"
              data-testid="responsive-header-1-2"
            >
              table header cell
            </div>
            <div
              class="flex-2"
              data-testid="responsive-data-1-2"
            >
              some more data
            </div>
          </div>
        </td>
      </tr>
      <tr
        class="border-base border-x border-y tablet-lg:border-0"
      >
        <td
          class="tablet-lg:display-table-cell border-base order-1 tablet-lg:order-0 display-block"
        >
          <div
            class="display-flex flex-align-center"
          >
            <div
              class="tablet-lg:display-none flex-1 text-bold"
              data-testid="responsive-header-2-0"
            >
              hi
            </div>
            <div
              class="flex-2"
              data-testid="responsive-data-2-0"
            >
              hi from row three
            </div>
          </div>
        </td>
        <td
          class="tablet-lg:display-table-cell border-base order-0 tablet-lg:order-1 display-block"
        >
          <div
            class="display-flex flex-align-center"
          >
            <div
              class="tablet-lg:display-none flex-1 text-bold"
              data-testid="responsive-header-2-1"
            >
              a heading
            </div>
            <div
              class="flex-2"
              data-testid="responsive-data-2-1"
            >
              column two
            </div>
          </div>
        </td>
        <td
          class="tablet-lg:display-table-cell border-base order--1 tablet-lg:order-2 display-none"
        >
          <div
            class="display-flex flex-align-center"
          >
            <div
              class="tablet-lg:display-none flex-1 text-bold"
              data-testid="responsive-header-2-2"
            >
              table header cell
            </div>
            <div
              class="flex-2"
              data-testid="responsive-data-2-2"
            >
              even more data
            </div>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
